os/drivers/otp/otp.c

```diff
@@ -0,0 +1,240 @@
+#include <errno.h>
+#include <debug.h>
+
+#include <tinyara/kmalloc.h>
+#include <tinyara/otp.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+#define OTP_DEV_PATH "/dev/otp"
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+struct otp_upperhalf_s {
+	/* Saved binding to the lower-half OTP driver */
+	struct otp_lowerhalf_s *ou_lower; /* Arch-specific operations */
+	sem_t ou_exclsem;	/* Supports exclusive access to the device */
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+static int     otp_open(struct file *filep);
+static int     otp_close(struct file *filep);
+static ssize_t otp_read(struct file *filep, char *buffer, size_t buflen);
+static ssize_t otp_write(struct file *filep, const char *buffer, size_t buflen);
+static int     otp_ioctl(struct file *filep, int cmd, unsigned long arg);
+
+/****************************************************************************
+ * Private Variables
+ ****************************************************************************/
+static const struct file_operations g_otp_ops = {
+	otp_open,	/* open  */
+	otp_close,	/* close */
+	otp_read,	/* read  */
+	otp_write,	/* write */
+	NULL,		/* seek  */
+	otp_ioctl,	/* ioctl */
+#ifndef CONFIG_DISABLE_POLL
+	NULL	/* poll  */
+#endif
+};
+
+static struct otp_upperhalf_s *g_otp_dev = NULL;
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+
+/****************************************************************************
+ * Name: otp_write
+ *
+ * Description:
+ *    This function is called when you handle the write() API in
+ *    File system.
+ *
+ ****************************************************************************/
+static ssize_t otp_write(struct file *filep, const char *buffer,
+			  size_t buflen)
+{
+	// Application is not allowed to write to OTP
+	return -1;
+}
+
+/****************************************************************************
+ * Name: otp_read
+ *
+ * Description:
+ *    This function is automatically called when you handle the read() API in
+ *    File system.
+ *
+ ****************************************************************************/
+static ssize_t otp_read(struct file *filep, char *buffer,
+			 size_t buflen)
+{
+	FAR struct inode *inode = filep->f_inode;
+	FAR struct otp_upperhalf_s *priv = inode->i_private;
+	FAR struct otp_lowerhalf_s *lower = priv->ou_lower;
+
+	// setting address should be updated
+	int res = lower->ops->read(lower, 0, (int8_t *)buffer, &buflen);
```

Comment:
lower should return one of err number in errno.h
vfs will set errno with returned value of each file operation.
please see fs_read.c